drivers/mac/controller.go

```diff
@@ -0,0 +1,157 @@
+// +build darwin,amd64
+
+package mac
+
+import (
+	"github.com/google/uuid"
+	"github.com/murlokswarm/app"
+	"github.com/murlokswarm/app/internal/bridge"
+	"github.com/murlokswarm/app/internal/core"
+)
+
+// Controller implements the core.Elem struct
+type Controller struct {
+	core.Controller
+
+	id string
+
+	// Event handlers
+	onDpadChange   func(app.ControllerInput, float64, float64)
+	onButtonChange func(app.ControllerInput, float64, bool)
+	onConnected    func()
+	onDisconnected func()
+	onPause        func()
+	onClose        func()
+}
+
+// newController creates an *app.Controller from
+// the specified app.ControllerConfig
+func newController(cc app.ControllerConfig) *Controller {
+	c := &Controller{
+		id:             uuid.New().String(),
+		onDpadChange:   cc.OnDpadChange,
+		onButtonChange: cc.OnButtonChange,
+		onConnected:    cc.OnConnected,
+		onDisconnected: cc.OnDisconnected,
+		onPause:        cc.OnPause,
+		onClose:        cc.OnClose,
+	}
+
+	if err := driver.macRPC.Call("controller.New", nil, struct {
+		ID string
+	}{
+		ID: c.id,
+	}); err != nil {
+		c.SetErr(err)
+		return c
+	}
+
+	driver.elems.Put(c)
+	return c
+}
+
+// ID satistfies the app.Elem interface.
+func (c *Controller) ID() string {
+	return c.id
+}
+
+// Close satisfies the app.Controller interface
+func (c *Controller) Close() {
+	err := driver.macRPC.Call("controller.Close", nil, struct {
+		ID string
+	}{
+		ID: c.id,
+	})
+
+	c.SetErr(err)
+	driver.elems.Delete(c)
+}
+
+// Called when either the directional pad or thumbsticks' values are changed (controller.onDpadChange)
+// Available values for `button`:
+//		dpad, leftThumbstick, rightThumbstick
+func onControllerDpadChange(c *Controller, in map[string]interface{}) interface{} {
+	if c.onDpadChange != nil {
+		input := app.ControllerInput(in["Input"].(float64))
+		x := in["X"].(float64)
+		y := in["Y"].(float64)
+		c.onDpadChange(input, x, y)
+	}
+
+	return nil
+}
+
+// Called when one of the following buttons' values are changed (controller.onButtonChange)
+// Available values for `button`:
+//		buttonA, buttonB, buttonX, buttonY,
+//		leftShoulder, rightShoulder
+//		leftTrigger, rightTrigger
+func onControllerButtonChange(c *Controller, in map[string]interface{}) interface{} {
+	if c.onButtonChange != nil {
```

Comment:
here you can put everything in the function call:
```go
c.onButtonChange(
		app.ControllerInput(in["Input"].(float64)),
		in["Value"].(float64),
		in["Pressed"].(bool),
	)
```

It will avoid to create temporary variable.

Reply:
made them inline params 👍